execution/block-partitioner/src/sharded_block_partitioner/mod.rs

```diff
@@ -0,0 +1,685 @@
+            frozen_rw_set_with_index.push(rw_set_with_index);
+            rejected_txns_vec.push(rejected_txns);
+        }
+        (frozen_chunks, frozen_rw_set_with_index, rejected_txns_vec)
+    }
+
+    fn discard_txns_with_cross_shard_dependencies(
+        &self,
+        txns_to_partition: Vec<Vec<AnalyzedTransaction>>,
+        frozen_chunks: Arc<Vec<SubBlock>>,
+        frozen_rw_set_with_index: Arc<Vec<RWSetWithTxnIndex>>,
+    ) -> (
+        Vec<SubBlock>,
+        Vec<RWSetWithTxnIndex>,
+        Vec<Vec<AnalyzedTransaction>>,
+    ) {
+        let partition_block_msgs = txns_to_partition
+            .into_iter()
+            .map(|txns| {
+                DiscardCrossShardDepReq(DiscardTxnsWithCrossShardDep::new(
+                    txns,
+                    frozen_rw_set_with_index.clone(),
+                    frozen_chunks.clone(),
+                ))
+            })
+            .collect();
+        self.send_partition_msgs(partition_block_msgs);
+        self.collect_partition_block_response()
+    }
+
+    fn add_cross_shard_dependencies(
+        &self,
+        index_offset: usize,
+        remaining_txns_vec: Vec<Vec<AnalyzedTransaction>>,
+        frozen_chunks: Arc<Vec<SubBlock>>,
+        frozen_rw_set_with_index: Arc<Vec<RWSetWithTxnIndex>>,
+    ) -> (
+        Vec<SubBlock>,
+        Vec<RWSetWithTxnIndex>,
+        Vec<Vec<AnalyzedTransaction>>,
+    ) {
+        let mut index_offset = index_offset;
+        let partition_block_msgs = remaining_txns_vec
+            .into_iter()
+            .map(|remaining_txns| {
+                let remaining_txns_len = remaining_txns.len();
+                let partitioning_msg = AddCrossShardDepReq(AddTxnsWithCrossShardDep::new(
+                    remaining_txns,
+                    index_offset,
+                    frozen_chunks.clone(),
+                    frozen_rw_set_with_index.clone(),
+                ));
+                index_offset += remaining_txns_len;
+                partitioning_msg
+            })
+            .collect::<Vec<ControlMsg>>();
+        self.send_partition_msgs(partition_block_msgs);
+        self.collect_partition_block_response()
+    }
+
+    /// We repeatedly partition chunks, discarding a bunch of transactions with cross-shard dependencies. The set of discarded
+    /// transactions are used as candidate chunks in the next round. This process is repeated until num_partitioning_rounds.
+    /// The remaining transactions are then added to the chunks with cross-shard dependencies.
+    pub fn partition(
+        &self,
+        transactions: Vec<AnalyzedTransaction>,
+        num_partitioning_round: usize,
+    ) -> Vec<SubBlock> {
+        let total_txns = transactions.len();
+        if total_txns == 0 {
+            return vec![];
+        }
+
+        // First round, we filter all transactions with cross-shard dependencies
+        let mut txns_to_partition = self.reorder_txns_by_senders(transactions);
```

Comment:
So this is reordering and partitioning..

Reply:
Yes, the naming is a bit confusing - open to suggestion here.

Reply:
`partition_by_senders` sounds more right..